overlord/devicestate/devicestate.go

```diff
@@ -341,91 +341,226 @@ func extractDownloadInstallEdgesFromTs(ts *state.TaskSet) (firstDl, lastDl, firs
 	return firstDl, tasks[edgeTaskIndex], tasks[edgeTaskIndex+1], lastInst, nil
 }
 
+func isNotInstalled(err error) bool {
+	_, ok := err.(*snap.NotInstalledError)
+	return ok
+}
+
 func notInstalled(st *state.State, name string) (bool, error) {
 	_, err := snapstate.CurrentInfo(st, name)
-	_, isNotInstalled := err.(*snap.NotInstalledError)
-	if isNotInstalled {
+	if isNotInstalled(err) {
 		return true, nil
 	}
 	return false, err
 }
 
+func kernelChange(current, new *asserts.Model) (changed bool, newKernelChannel string) {
+	if new.Grade() == asserts.ModelGradeUnset {
+		if current.Kernel() != new.Kernel() {
+			return true, new.KernelTrack()
+		}
+		trackChange := current.KernelTrack() != new.KernelTrack()
+		if trackChange {
+			return true, new.KernelTrack()
+		}
+	} else {
+		if current.Kernel() != new.Kernel() {
+			return true, new.KernelSnap().DefaultChannel
+		}
+		defaultChannelChange := current.KernelSnap().DefaultChannel != new.KernelSnap().DefaultChannel
```

Comment:
thinking further, shouldn't this code actually check the Tracking channel instead of the old model for the UC20 case?

Reply:
Hmm added this, the test suite grew a bit again, there's more state to be mocked now.